util/compression.h

```diff
@@ -144,6 +147,390 @@ class ZSTDUncompressCachedData {
 
 namespace ROCKSDB_NAMESPACE {
 
+// ***********************************************************************
+// BEGIN future compression customization interface
+// ***********************************************************************
+
+// TODO: alias/adapt for compression
+struct FilterBuildingContext;
+
+// A Compressor represents a very specific but potentially adapting strategy for
+// compressing blocks, including the relevant algorithm(s), options, dictionary,
+// etc. as applicable--every input except the sequence of bytes to compress.
+// Compressor is generally thread-safe so can be shared by multiple threads. (It
+// might make sense to convert unique_ptr<Compressor> to
+// shared_ptr<Compressor>.) A Compressor for data files is expected to be used
+// for just one file, so that compression strategy can be explicitly
+// reconsidered for each new file. However, a Compressor for in-memory use could
+// live indefinitely.
+//
+// If a single thread is doing many compressions under the same strategy, it
+// should request a WorkingArea that will in some cases make repeated
+// compression in a single thread more efficient. Unlike the rest of Compressor,
+// each WorkingArea can only be used by one thread at a time. WorkingAreas can
+// have pre-allocated space and/or data structures, and/or thread-local
+// statistics that are later incorporated into shared statistics objects.
+//
+// The Compressor marks each block with a CompressionType to guide
+// decompression. However, the compression dictionary (or whether there is one
+// associated) is determined at Compressor creation time, though the process of
+// getting a Compressor with a dictionary starts with a Compressor without
+// dictionary (which will often be relevant alongside); see relevant functions.
+// If the Compressor wants to decide block-by-block whether to apply the
+// configured dictionary, that would need to be encoded in CompressionType or
+// the compressed output.
+class Compressor {
+ public:
+  Compressor() = default;
+  virtual ~Compressor() = default;
+
+  // Returns the max total bytes of for all sampled blocks for creating the data
+  // dictionary, or zero indicating dictionary compression should not be
+  // used/configured. This will typically be called after
+  // CompressionManager::GetCompressor() to see if samples should be accumulated
+  // and passed to MaybeCloneSpecialized().
+  virtual size_t GetMaxSampleSizeIfWantDict(CacheEntryRole block_type) const {
+    // Default implementation: no dictionary
```

Comment:
I guess this means no explicit sample passing into Compressor from RocksDB but does that equal to no dictionary compression? Can Compressor impl internally sample data through CompressBlock() call?

Reply:
To create a Compressor with a dictionary, the caller uses MaybeCloneSpecialized() below, likely after calling this function to let the Compressor decide whether it prefers to use dictionary compression. I guess I didn't finish filling out that functions documentation.

It might seem like an odd design but there are some reasons for it. (I'll try to provide more background)